File: src/outage-tracker.ts
import { NetworkMetric, OutageEvent } from './types';

export class OutageTracker {
  // How stale an ongoing outage can be on load before we close it instead of resuming.
  // We can't claim downtime that occurred while the monitor wasn't running.
  private static readonly STALE_OUTAGE_MS = 5 * 60 * 1000;

  private currentOutage: OutageEvent | null = null;
  private outages: OutageEvent[] = [];
  private outageThreshold = {
    packetLoss: 50,
    consecutiveFailures: 2
  };
  private consecutiveFailures = 0;
  private firstFailureMetric: NetworkMetric | null = null;

  processMetric(metric: NetworkMetric): OutageEvent | null {
    const isOutage = this.isOutageCondition(metric);
    metric.isOutage = isOutage;

    if (isOutage) {
      this.consecutiveFailures++;
      if (!this.firstFailureMetric) {
        this.firstFailureMetric = metric;
      }

      if (this.currentOutage) {
        this.currentOutage.lastUpdateTime = metric.timestamp;
      } else if (this.consecutiveFailures >= this.outageThreshold.consecutiveFailures) {
        this.currentOutage = this.startOutage(this.firstFailureMetric, metric);
        return this.currentOutage;
      }
    } else {
      this.firstFailureMetric = null;
      if (this.currentOutage) {
        const endedOutage = this.endOutage(metric);
        this.consecutiveFailures = 0;
        return endedOutage;
      }
      this.consecutiveFailures = 0;
    }

    return null;
  }

  private isOutageCondition(metric: NetworkMetric): boolean {
    const hasHighPacketLoss = metric.ping.packetLoss >= this.outageThreshold.packetLoss;
    const hasDnsFailure = !metric.dns.success;
    const hasNoConnectivity = metric.ping.packetLoss === 100;

    return hasNoConnectivity || (hasHighPacketLoss && hasDnsFailure);
  }

  private startOutage(firstFailure: NetworkMetric, latestFailure: NetworkMetric): OutageEvent {
    const outage: OutageEvent = {
      id: `outage-${firstFailure.timestamp.getTime()}`,
      startTime: firstFailure.timestamp,
      lastUpdateTime: latestFailure.timestamp,
      type: firstFailure.ping.packetLoss === 100 ? 'connectivity' : 'partial',
      metrics: {
        packetLoss: firstFailure.ping.packetLoss,
        dnsFailure: !firstFailure.dns.success
      }
    };

    this.outages.push(outage);
    return outage;
  }

  private endOutage(metric: NetworkMetric): OutageEvent {
    if (!this.currentOutage) {
      throw new Error('No current outage to end');
    }

    this.currentOutage.endTime = metric.timestamp;
    this.currentOutage.duration = 
      metric.timestamp.getTime() - this.currentOutage.startTime.getTime();

    const endedOutage = this.currentOutage;
    this.currentOutage = null;
    return endedOutage;
  }

  getOutages(since?: Date): OutageEvent[] {
    if (!since) return [...this.outages];
    
    return this.outages.filter(o => o.startTime >= since);
  }

  getCurrentOutage(): OutageEvent | null {
    return this.currentOutage;
  }

  clearOutages(): void {
    this.outages = [];
    this.currentOutage = null;
    this.consecutiveFailures = 0;
    this.firstFailureMetric = null;
  }

  loadOutages(outages: OutageEvent[]): void {
    this.outages = outages.map(o => ({
      ...o,
      startTime: new Date(o.startTime),
      endTime: o.endTime ? new Date(o.endTime) : undefined,
      lastUpdateTime: o.lastUpdateTime ? new Date(o.lastUpdateTime) : undefined
    }));

    const ongoingOutage = this.outages.find(o => !o.endTime);
    if (!ongoingOutage) return;

    const lastActivity = ongoingOutage.lastUpdateTime ?? ongoingOutage.startTime;
    const staleness = Date.now() - lastActivity.getTime();

    if (staleness > OutageTracker.STALE_OUTAGE_MS) {
      // Monitor wasn't running for too long to credibly extend this outage; close it
      // at the last observed activity so stats don't double-count the offline gap.
      ongoingOutage.endTime = lastActivity;
      ongoingOutage.duration = lastActivity.getTime() - ongoingOutage.startTime.getTime();
    } else {
      this.currentOutage = ongoingOutage;
    }
  }
}

File: mac/Netmon/Engine/OutageTracker.swift
import Foundation

/// Detects connectivity outages from a stream of NetworkMetric samples.
/// Mirrors the TypeScript OutageTracker with the recent fixes:
///   * outage startTime is the *first* failing metric, not the second
///   * lastUpdateTime is tracked while ongoing so a restart can tell how
///     recently we observed activity
///   * stale ongoing outages from a prior session are closed at load time
///     so they don't inflate stats with `Date.now() - startTime`.
struct OutageTracker {
    struct ProcessResult {
        /// State-changing event (open/close). `nil` when nothing transitioned.
        var event: OutageEvent?
        /// The currently-ongoing outage with its updated lastUpdateTime, if any.
        /// Callers should persist this if `event == nil && metric.isOutage`
        /// to keep lastUpdateTime fresh across restarts.
        var ongoing: OutageEvent?
    }

    /// How stale an ongoing outage can be on load before we close it instead
    /// of resuming. We can't claim downtime that occurred while the monitor
    /// wasn't running.
    static let staleOutageInterval: TimeInterval = 5 * 60

    private let packetLossThreshold: Double = 50
    private let consecutiveFailuresThreshold = 2

    private(set) var outages: [OutageEvent] = []
    private(set) var currentOutage: OutageEvent?

    private var consecutiveFailures = 0
    private var firstFailureMetric: NetworkMetric?

    mutating func processMetric(_ metricIn: NetworkMetric) -> ProcessResult {
        var metric = metricIn
        let isOutage = isOutageCondition(metric)
        metric.isOutage = isOutage

        if isOutage {
            consecutiveFailures += 1
            if firstFailureMetric == nil { firstFailureMetric = metric }

            if var current = currentOutage {
                current.lastUpdateTime = metric.timestamp
                currentOutage = current
                replaceInOutages(current)
                return ProcessResult(event: nil, ongoing: current)
            }

            if consecutiveFailures >= consecutiveFailuresThreshold,
               let first = firstFailureMetric {
                let new = startOutage(firstFailure: first, latestFailure: metric)
                currentOutage = new
                outages.append(new)
                return ProcessResult(event: new, ongoing: new)
            }

            return ProcessResult(event: nil, ongoing: nil)
        }

        firstFailureMetric = nil
        if var current = currentOutage {
            current.endTime = metric.timestamp
            current.durationMs = metric.timestamp.timeIntervalSince(current.startTime) * 1000
            current.lastUpdateTime = metric.timestamp
            currentOutage = nil
            consecutiveFailures = 0
            replaceInOutages(current)
            return ProcessResult(event: current, ongoing: nil)
        }
        consecutiveFailures = 0
        return ProcessResult(event: nil, ongoing: nil)
    }

    mutating func loadPersisted(_ persisted: [OutageEvent], now: Date = .now) {
        outages = persisted

        guard let idx = outages.firstIndex(where: { $0.endTime == nil }) else { return }
        var ongoing = outages[idx]
        let lastActivity = ongoing.lastUpdateTime ?? ongoing.startTime
        let staleness = now.timeIntervalSince(lastActivity)

        if staleness > Self.staleOutageInterval {
            ongoing.endTime = lastActivity
            ongoing.durationMs = lastActivity.timeIntervalSince(ongoing.startTime) * 1000
            outages[idx] = ongoing
        } else {
            currentOutage = ongoing
        }
    }

    private func isOutageCondition(_ metric: NetworkMetric) -> Bool {
        let hasHighPacketLoss = metric.pingPacketLoss >= packetLossThreshold
        let hasDNSFailure = !metric.dnsSuccess
        let hasNoConnectivity = metric.pingPacketLoss == 100
        return hasNoConnectivity || (hasHighPacketLoss && hasDNSFailure)
    }

    private func startOutage(firstFailure: NetworkMetric, latestFailure: NetworkMetric) -> OutageEvent {
        OutageEvent(
            id: "outage-\(Int(firstFailure.timestamp.timeIntervalSince1970 * 1000))",
            startTime: firstFailure.timestamp,
            endTime: nil,
            durationMs: nil,
            lastUpdateTime: latestFailure.timestamp,
            type: firstFailure.pingPacketLoss == 100 ? .connectivity : .partial,
            startPacketLoss: firstFailure.pingPacketLoss,
            startDNSFailure: !firstFailure.dnsSuccess
        )
    }

    private mutating func replaceInOutages(_ updated: OutageEvent) {
        if let i = outages.firstIndex(where: { $0.id == updated.id }) {
            outages[i] = updated
        }
    }
}
